### PR TITLE
Plane wrapping

### PR DESCRIPTION
**Pull request by bru345**
Decal system by buffer geometry position manipulation.

How to:
1. Pick up gun and fire at any mesh.
2. See decal appear

Known issues: 
1. Decal system, like DecalGeometry.js, has issues with certain meshes. Can appear inside some meshes, although it's fine on most.

Planned Optimisation: 
1. Moving Decal in its own repo.
2. Expose values (size, texture image, etc) so other systems can easily utilise it
3. Manage decals in scene either by Objectpool or use Three Instance mesh - https://threejs.org/docs/#api/en/objects/InstancedMesh 
4. Use promise instead of setTimeout.
5. z-fighting issues (could be fixed shader side)

Notes: 
You can change the decal image by placing a new jpg or png in the root folder and updating the variable 'decalTextureName' with the file name.

You can remove the green debug cubes by flipping the bool 'debugDecalVertPos' to false.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -157,13 +169,82 @@ export default () => {
...
+            ));
+
+            scene.add(plane);
+            plane.updateMatrix();
+
+            let positions = planeGeo.attributes.position.array;
+            let ptCout = positions.length;
+
+            // Decal vertex manipulation
+            setTimeout(() => {
```
Comment (avaer) [inline index.js:198]: Why `setTimeout`?

File: index.js [hunk 2]
```diff
@@ -157,13 +169,82 @@ export default () => {
...
+
+            scene.add(plane);
+            plane.updateMatrix();
+
+            let positions = planeGeo.attributes.position.array;
+            let ptCout = positions.length;
+
+            // Decal vertex manipulation
+            setTimeout(() => {  
+              if (planeGeo instanceof THREE.BufferGeometry)
```
Comment (avaer) [inline index.js:199]: Why would this be false?

**Review by avaer — CHANGES_REQUESTED** [review]
Some code questions, but the bigger issue is that this is creating a new untracked plane on each hit.

Imo the planes should all be accumulated into a single mesh, with a draw range limited BufferGeometry. Then you can copy in new PlaneBufferGeometry attributes transformed with the correct matrix, pasted into the BufferGeometry at the right index, with the update range and draw range set accordingly.

Futhermore, when we break this out in to a separate decal app, it should clean up after itself and not leave the meshes lying around. The `useCleanup` hook is for that.

**Comment by avaer** [comment]
The reason for the above is performance; all of the decals can and should be a single draw call.

**Comment by avaer** [comment]
https://threejs.org/docs/#api/en/core/BufferGeometry.drawRange

File: index.js [hunk 3]
```diff
@@ -157,13 +169,82 @@ export default () => {
...
           const result = physics.raycast(gunApp.position, gunApp.quaternion.clone().multiply(z180Quaternion));
           if (result) {
-            explosionApp.position.fromArray(result.point);
+            // Decal creation
             const normal = new THREE.Vector3().fromArray(result.normal);
+            const planeGeo = new THREE.PlaneBufferGeometry(0.5, 0.5, 8, 8)
+            let plane = new THREE.Mesh( planeGeo, decalMaterial);
+            plane.name = "DecalPlane"
+            const newPointVec = new THREE.Vector3().fromArray(result.point);
+            const modiPoint = newPointVec.add(new Vector3(0, normal.y /20 ,0));
```
Comment (avaer) [inline index.js:183]: This factor looks like it should be a constant to me.

**Comment by avaer** [comment]
(the main issue is point 3 in the description)

> Manage decals in scene either by Objectpool or use Three Instance mesh - https://threejs.org/docs/#api/en/objects/InstancedMesh

Though I don't know if this will make sense with instanced mesh. All of the geos are actually unique, so there isn't much to instance, it could just be a vertex accumulation.

**Comment by avaer** [comment]
Also @skaljac is working on a way to do multiple raycasts in a single call. When that is done it is worth using here, again for performance.

File: index.js (hunk 2 above)
Comment (bru345) [reply index.js:199]: because of execution of order, sometimes planeGeo would be undefined.  It's redundant and should be removed.

**Comment by avaer** [comment]
This looks ok for now, will merge and continue optimization work in pistol master.

**Pull request merged**